application/src/main/java/bisq/application/DefaultApplicationService.java

```diff
@@ -271,4 +274,30 @@ private boolean isRegtestRun() {
                 .build();
         return Optional.of(walletConfig);
     }
+
+    private CompletableFuture<Boolean> setState(State newState) {
+        checkArgument(state.get().ordinal() < newState.ordinal(),
+                "New state %s must have a higher ordinal as the current state %s", newState, state.get());
+        state.set(newState);
+        log.info("New state {}", newState);
+        runAsync(() -> listeners.forEach(e -> e.onStateChanged(newState)), DISPATCHER);
+
+        return CompletableFuture.completedFuture(true);
+    }
+
+    private CompletableFuture<Boolean> setStateAfter(CompletableFuture<Boolean> initTask, State stateAfter) {
+        return initTask.thenCompose(result -> setState(stateAfter));
```

Comment:
I think the setState could be handled in a whenComplete instead of staring a new CompletableFuture with thenCompose